File: rb/MemesioContentCreation_sdk.rb
# MemesioContentCreation SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'MemesioContentCreation_types'


class MemesioContentCreationSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = MemesioContentCreationUtility.new
    @_utility = utility

    config = MemesioContentCreationConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = MemesioContentCreationHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = MemesioContentCreationHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, MemesioContentCreationFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    MemesioContentCreationUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = MemesioContentCreationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = MemesioContentCreationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = MemesioContentCreationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = MemesioContentCreationSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue MemesioContentCreationError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = MemesioContentCreationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = MemesioContentCreationHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.Agent.list / client.Agent.load({ "id" => ... })
  def Agent(data = nil)
    require_relative 'entity/agent_entity'
    AgentEntity.new(self, data)
  end


  # Canonical facade: client.AgentInfra.list / client.AgentInfra.load({ "id" => ... })
  def AgentInfra(data = nil)
    require_relative 'entity/agent_infra_entity'
    AgentInfraEntity.new(self, data)
  end


  # Canonical facade: client.AiCaption.list / client.AiCaption.load({ "id" => ... })
  def AiCaption(data = nil)
    require_relative 'entity/ai_caption_entity'
    AiCaptionEntity.new(self, data)
  end


  # Canonical facade: client.AiJob.list / client.AiJob.load({ "id" => ... })
  def AiJob(data = nil)
    require_relative 'entity/ai_job_entity'
    AiJobEntity.new(self, data)
  end


  # Canonical facade: client.AiMemeGenerationSucceeded.list / client.AiMemeGenerationSucceeded.load({ "id" => ... })
  def AiMemeGenerationSucceeded(data = nil)
    require_relative 'entity/ai_meme_generation_succeeded_entity'
    AiMemeGenerationSucceededEntity.new(self, data)
  end


  # Canonical facade: client.AiProvider.list / client.AiProvider.load({ "id" => ... })
  def AiProvider(data = nil)
    require_relative 'entity/ai_provider_entity'
    AiProviderEntity.new(self, data)
  end


  # Canonical facade: client.Analytics.list / client.Analytics.load({ "id" => ... })
  def Analytics(data = nil)
    require_relative 'entity/analytics_entity'
    AnalyticsEntity.new(self, data)
  end


  # Canonical facade: client.Auth.list / client.Auth.load({ "id" => ... })
  def Auth(data = nil)
    require_relative 'entity/auth_entity'
    AuthEntity.new(self, data)
  end


  # Canonical facade: client.Billing.list / client.Billing.load({ "id" => ... })
  def Billing(data = nil)
    require_relative 'entity/billing_entity'
    BillingEntity.new(self, data)
  end


  # Canonical facade: client.Collaboration.list / client.Collaboration.load({ "id" => ... })
  def Collaboration(data = nil)
    require_relative 'entity/collaboration_entity'
    CollaborationEntity.new(self, data)
  end


  # Canonical facade: client.Compliance.list / client.Compliance.load({ "id" => ... })
  def Compliance(data = nil)
    require_relative 'entity/compliance_entity'
    ComplianceEntity.new(self, data)
  end


  # Canonical facade: client.CreateMeme.list / client.CreateMeme.load({ "id" => ... })
  def CreateMeme(data = nil)
    require_relative 'entity/create_meme_entity'
    CreateMemeEntity.new(self, data)
  end


  # Canonical facade: client.DeveloperApi.list / client.DeveloperApi.load({ "id" => ... })
  def DeveloperApi(data = nil)
    require_relative 'entity/developer_api_entity'
    DeveloperApiEntity.new(self, data)
  end


  # Canonical facade: client.FreeCaptionMemeSuccess.list / client.FreeCaptionMemeSuccess.load({ "id" => ... })
  def FreeCaptionMemeSuccess(data = nil)
    require_relative 'entity/free_caption_meme_success_entity'
    FreeCaptionMemeSuccessEntity.new(self, data)
  end


  # Canonical facade: client.FreeTemplateSearch.list / client.FreeTemplateSearch.load({ "id" => ... })
  def FreeTemplateSearch(data = nil)
    require_relative 'entity/free_template_search_entity'
    FreeTemplateSearchEntity.new(self, data)
  end


  # Canonical facade: client.Generate.list / client.Generate.load({ "id" => ... })
  def Generate(data = nil)
    require_relative 'entity/generate_entity'
    GenerateEntity.new(self, data)
  end


  # Canonical facade: client.Growth.list / client.Growth.load({ "id" => ... })
  def Growth(data = nil)
    require_relative 'entity/growth_entity'
    GrowthEntity.new(self, data)
  end


  # Canonical facade: client.ListMeme.list / client.ListMeme.load({ "id" => ... })
  def ListMeme(data = nil)
    require_relative 'entity/list_meme_entity'
    ListMemeEntity.new(self, data)
  end


  # Canonical facade: client.Media.list / client.Media.load({ "id" => ... })
  def Media(data = nil)
    require_relative 'entity/media_entity'
    MediaEntity.new(self, data)
  end


  # Canonical facade: client.Meme.list / client.Meme.load({ "id" => ... })
  def Meme(data = nil)
    require_relative 'entity/meme_entity'
    MemeEntity.new(self, data)
  end


  # Canonical facade: client.PublicTemplateMediaItem.list / client.PublicTemplateMediaItem.load({ "id" => ... })
  def PublicTemplateMediaItem(data = nil)
    require_relative 'entity/public_template_media_item_entity'
    PublicTemplateMediaItemEntity.new(self, data)
  end


  # Canonical facade: client.StandaloneAgentBootstrap.list / client.StandaloneAgentBootstrap.load({ "id" => ... })
  def StandaloneAgentBootstrap(data = nil)
    require_relative 'entity/standalone_agent_bootstrap_entity'
    StandaloneAgentBootstrapEntity.new(self, data)
  end


  # Canonical facade: client.Template.list / client.Template.load({ "id" => ... })
  def Template(data = nil)
    require_relative 'entity/template_entity'
    TemplateEntity.new(self, data)
  end


  # Canonical facade: client.TemplateSearch.list / client.TemplateSearch.load({ "id" => ... })
  def TemplateSearch(data = nil)
    require_relative 'entity/template_search_entity'
    TemplateSearchEntity.new(self, data)
  end


  # Canonical facade: client.TrendAlert.list / client.TrendAlert.load({ "id" => ... })
  def TrendAlert(data = nil)
    require_relative 'entity/trend_alert_entity'
    TrendAlertEntity.new(self, data)
  end


  # Canonical facade: client.UploadCaptionMemeSuccess.list / client.UploadCaptionMemeSuccess.load({ "id" => ... })
  def UploadCaptionMemeSuccess(data = nil)
    require_relative 'entity/upload_caption_meme_success_entity'
    UploadCaptionMemeSuccessEntity.new(self, data)
  end


  # Canonical facade: client.Video.list / client.Video.load({ "id" => ... })
  def Video(data = nil)
    require_relative 'entity/video_entity'
    VideoEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = MemesioContentCreationSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/memesiocontentcreation_sdk.py
# MemesioContentCreation SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import MemesioContentCreationUtility
from core.spec import MemesioContentCreationSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import MemesioContentCreationBaseFeature
from features import _make_feature


class MemesioContentCreationSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = MemesioContentCreationUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return MemesioContentCreationUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = MemesioContentCreationSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Agent(self, data=None) -> "AgentEntity":
        """Entity factory: client.Agent().list() / client.Agent().load({"id": ...})."""
        from entity.agent_entity import AgentEntity
        return AgentEntity(self, data)


    def AgentInfra(self, data=None) -> "AgentInfraEntity":
        """Entity factory: client.AgentInfra().list() / client.AgentInfra().load({"id": ...})."""
        from entity.agent_infra_entity import AgentInfraEntity
        return AgentInfraEntity(self, data)


    def AiCaption(self, data=None) -> "AiCaptionEntity":
        """Entity factory: client.AiCaption().list() / client.AiCaption().load({"id": ...})."""
        from entity.ai_caption_entity import AiCaptionEntity
        return AiCaptionEntity(self, data)


    def AiJob(self, data=None) -> "AiJobEntity":
        """Entity factory: client.AiJob().list() / client.AiJob().load({"id": ...})."""
        from entity.ai_job_entity import AiJobEntity
        return AiJobEntity(self, data)


    def AiMemeGenerationSucceeded(self, data=None) -> "AiMemeGenerationSucceededEntity":
        """Entity factory: client.AiMemeGenerationSucceeded().list() / client.AiMemeGenerationSucceeded().load({"id": ...})."""
        from entity.ai_meme_generation_succeeded_entity import AiMemeGenerationSucceededEntity
        return AiMemeGenerationSucceededEntity(self, data)


    def AiProvider(self, data=None) -> "AiProviderEntity":
        """Entity factory: client.AiProvider().list() / client.AiProvider().load({"id": ...})."""
        from entity.ai_provider_entity import AiProviderEntity
        return AiProviderEntity(self, data)


    def Analytics(self, data=None) -> "AnalyticsEntity":
        """Entity factory: client.Analytics().list() / client.Analytics().load({"id": ...})."""
        from entity.analytics_entity import AnalyticsEntity
        return AnalyticsEntity(self, data)


    def Auth(self, data=None) -> "AuthEntity":
        """Entity factory: client.Auth().list() / client.Auth().load({"id": ...})."""
        from entity.auth_entity import AuthEntity
        return AuthEntity(self, data)


    def Billing(self, data=None) -> "BillingEntity":
        """Entity factory: client.Billing().list() / client.Billing().load({"id": ...})."""
        from entity.billing_entity import BillingEntity
        return BillingEntity(self, data)


    def Collaboration(self, data=None) -> "CollaborationEntity":
        """Entity factory: client.Collaboration().list() / client.Collaboration().load({"id": ...})."""
        from entity.collaboration_entity import CollaborationEntity
        return CollaborationEntity(self, data)


    def Compliance(self, data=None) -> "ComplianceEntity":
        """Entity factory: client.Compliance().list() / client.Compliance().load({"id": ...})."""
        from entity.compliance_entity import ComplianceEntity
        return ComplianceEntity(self, data)


    def CreateMeme(self, data=None) -> "CreateMemeEntity":
        """Entity factory: client.CreateMeme().list() / client.CreateMeme().load({"id": ...})."""
        from entity.create_meme_entity import CreateMemeEntity
        return CreateMemeEntity(self, data)


    def DeveloperApi(self, data=None) -> "DeveloperApiEntity":
        """Entity factory: client.DeveloperApi().list() / client.DeveloperApi().load({"id": ...})."""
        from entity.developer_api_entity import DeveloperApiEntity
        return DeveloperApiEntity(self, data)


    def FreeCaptionMemeSuccess(self, data=None) -> "FreeCaptionMemeSuccessEntity":
        """Entity factory: client.FreeCaptionMemeSuccess().list() / client.FreeCaptionMemeSuccess().load({"id": ...})."""
        from entity.free_caption_meme_success_entity import FreeCaptionMemeSuccessEntity
        return FreeCaptionMemeSuccessEntity(self, data)


    def FreeTemplateSearch(self, data=None) -> "FreeTemplateSearchEntity":
        """Entity factory: client.FreeTemplateSearch().list() / client.FreeTemplateSearch().load({"id": ...})."""
        from entity.free_template_search_entity import FreeTemplateSearchEntity
        return FreeTemplateSearchEntity(self, data)


    def Generate(self, data=None) -> "GenerateEntity":
        """Entity factory: client.Generate().list() / client.Generate().load({"id": ...})."""
        from entity.generate_entity import GenerateEntity
        return GenerateEntity(self, data)


    def Growth(self, data=None) -> "GrowthEntity":
        """Entity factory: client.Growth().list() / client.Growth().load({"id": ...})."""
        from entity.growth_entity import GrowthEntity
        return GrowthEntity(self, data)


    def ListMeme(self, data=None) -> "ListMemeEntity":
        """Entity factory: client.ListMeme().list() / client.ListMeme().load({"id": ...})."""
        from entity.list_meme_entity import ListMemeEntity
        return ListMemeEntity(self, data)


    def Media(self, data=None) -> "MediaEntity":
        """Entity factory: client.Media().list() / client.Media().load({"id": ...})."""
        from entity.media_entity import MediaEntity
        return MediaEntity(self, data)


    def Meme(self, data=None) -> "MemeEntity":
        """Entity factory: client.Meme().list() / client.Meme().load({"id": ...})."""
        from entity.meme_entity import MemeEntity
        return MemeEntity(self, data)


    def PublicTemplateMediaItem(self, data=None) -> "PublicTemplateMediaItemEntity":
        """Entity factory: client.PublicTemplateMediaItem().list() / client.PublicTemplateMediaItem().load({"id": ...})."""
        from entity.public_template_media_item_entity import PublicTemplateMediaItemEntity
        return PublicTemplateMediaItemEntity(self, data)


    def StandaloneAgentBootstrap(self, data=None) -> "StandaloneAgentBootstrapEntity":
        """Entity factory: client.StandaloneAgentBootstrap().list() / client.StandaloneAgentBootstrap().load({"id": ...})."""
        from entity.standalone_agent_bootstrap_entity import StandaloneAgentBootstrapEntity
        return StandaloneAgentBootstrapEntity(self, data)


    def Template(self, data=None) -> "TemplateEntity":
        """Entity factory: client.Template().list() / client.Template().load({"id": ...})."""
        from entity.template_entity import TemplateEntity
        return TemplateEntity(self, data)


    def TemplateSearch(self, data=None) -> "TemplateSearchEntity":
        """Entity factory: client.TemplateSearch().list() / client.TemplateSearch().load({"id": ...})."""
        from entity.template_search_entity import TemplateSearchEntity
        return TemplateSearchEntity(self, data)


    def TrendAlert(self, data=None) -> "TrendAlertEntity":
        """Entity factory: client.TrendAlert().list() / client.TrendAlert().load({"id": ...})."""
        from entity.trend_alert_entity import TrendAlertEntity
        return TrendAlertEntity(self, data)


    def UploadCaptionMemeSuccess(self, data=None) -> "UploadCaptionMemeSuccessEntity":
        """Entity factory: client.UploadCaptionMemeSuccess().list() / client.UploadCaptionMemeSuccess().load({"id": ...})."""
        from entity.upload_caption_meme_success_entity import UploadCaptionMemeSuccessEntity
        return UploadCaptionMemeSuccessEntity(self, data)


    def Video(self, data=None) -> "VideoEntity":
        """Entity factory: client.Video().list() / client.Video().load({"id": ...})."""
        from entity.video_entity import VideoEntity
        return VideoEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "MemesioContentCreationSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.agent_entity import AgentEntity
    from entity.agent_infra_entity import AgentInfraEntity
    from entity.ai_caption_entity import AiCaptionEntity
    from entity.ai_job_entity import AiJobEntity
    from entity.ai_meme_generation_succeeded_entity import AiMemeGenerationSucceededEntity
    from entity.ai_provider_entity import AiProviderEntity
    from entity.analytics_entity import AnalyticsEntity
    from entity.auth_entity import AuthEntity
    from entity.billing_entity import BillingEntity
    from entity.collaboration_entity import CollaborationEntity
    from entity.compliance_entity import ComplianceEntity
    from entity.create_meme_entity import CreateMemeEntity
    from entity.developer_api_entity import DeveloperApiEntity
    from entity.free_caption_meme_success_entity import FreeCaptionMemeSuccessEntity
    from entity.free_template_search_entity import FreeTemplateSearchEntity
    from entity.generate_entity import GenerateEntity
    from entity.growth_entity import GrowthEntity
    from entity.list_meme_entity import ListMemeEntity
    from entity.media_entity import MediaEntity
    from entity.meme_entity import MemeEntity
    from entity.public_template_media_item_entity import PublicTemplateMediaItemEntity
    from entity.standalone_agent_bootstrap_entity import StandaloneAgentBootstrapEntity
    from entity.template_entity import TemplateEntity
    from entity.template_search_entity import TemplateSearchEntity
    from entity.trend_alert_entity import TrendAlertEntity
    from entity.upload_caption_meme_success_entity import UploadCaptionMemeSuccessEntity
    from entity.video_entity import VideoEntity
